
# Dataset Information:

- The dataset is acquired from `The Energy Institute`, you can load this data from the following link at https://www.energyinst.org/statistical-review/resources-and-data-downloads.

- The data offers details data on global energy production, consumption, and emissions, from various energy sources such as oil, natural gas, coal, nuclear, hydroelectric, and renewables over spans multiple years.


In [1]:
import mysql.connector
from mysql.connector import errorcode
import csv
import sqlalchemy as sq
# In the lines below and fill in any relevant details, such as if you set up a different user or password
# Importing keys
import config

PASSWORD=config.sql_key
myconnection = mysql.connector.connect(user='root', password=PASSWORD,
                                host='127.0.0.1', database='data604',port=3306)


In [3]:
import pandas as pd
import numpy as np
df = pd.read_csv('Ren_gen_energy_data.csv')

In [6]:
df.head()

Country   Wind_2022  Solar_2022  Hydro_2022  OtherRenewables_2022   
0     Canada   38.376387    7.050284  397.725241              9.123262  \
1     Mexico   20.528750   20.342040   35.558850              6.696620   
2         US  438.684200  207.150505  251.273544             71.509097   
3  Argentina   14.164993    2.940079   22.769590              2.352800   
4     Brazil   81.631569   30.126462  427.113549             55.033215   

    Wind_2023  Solar_2023  Hydro_2023  OtherRenewables_2023  
0   38.936293    7.641602  364.196497             10.384484  
1   21.650982   27.143349   20.399428              7.099758  
2  429.530709  240.525253  236.321296             67.290103  
3   14.475425    3.272582   29.898193              2.275717  
4   95.508935   51.482100  428.653200             55.832700

In [7]:
# Creating a cursor 
create_cursor = myconnection.cursor()

# CREATE TABLE
create_statement = '''
create table renewable_energy (
    Country varchar(50) NOT NULL,
    Wind_2022 float,
    Solar_2022 float,
    Hydro_2022 float,
    OtherRenewables_2022 float,
    Wind_2023 float,
    Solar_2023 float,
    Hydro_2023 float,
    OtherRenewables_2023 float
);
'''

try:
    create_cursor.execute(create_statement)
except mysql.connector.Error as err:
    if err.errno == errorcode.ER_TABLE_EXISTS_ERROR:
        print("Table already exists.")
    else:
        print(err.msg)
else:
    print("Table created successfully!")

create_cursor.close()

Table already exists.


True

In [8]:
# Column names
PATH = "Correctly_Aligned_Renewable_Energy_Data.csv"
with open(PATH, 'r') as file:
    reader = csv.reader(file)
    column_names = next(reader)

print(column_names)


['Country', 'Wind_2022', 'Solar_2022', 'Hydro_2022', 'OtherRenewables_2022', 'Wind_2023', 'Solar_2023', 'Hydro_2023', 'OtherRenewables_2023']


In [ ]:
# Inserting data into the table
insert_cursor = myconnection.cursor()

# Prepare column names and values
columnString = "`,`".join([str(col) for col in column_names])
# print(columnString)

with open(PATH, mode='r') as csv_file:
    csv_reader = csv.reader(csv_file)
    next(csv_reader)
    for row in csv_reader:
        # print(row)      
        insertCommand = f"INSERT INTO `renewable_energy` (`{columnString}`) VALUES (" + "%s," * (len(row) - 1) + "%s)"
        # print(insertCommand)
        insert_cursor.execute(insertCommand, tuple(row))
        
print("Data inserted successfully!")

# Commiting the changes
myconnection.commit()
insert_cursor.close()

Data inserted successfully!


# Queries

#### Understanding the dataset:

All numerical values in the dataset are represented in TWh (Terawatt-Hours).
A **Terawatt-Hour (TWh)** is a unit of energy equivalent to `one trillion watts` $(10^{12} watts)$ consumed or generated for one hour.
- One TWh could power:
    - The entire state of California for about 1.5 weeks.
    - 100 million homes for an hour.
    - The world’s data centers (which power the internet) for about 15 days.
- In 2021, Canada produced *625.7 terawatt-hours (TWh)* of electricity. 
- More than half of the electricity in Canada (60%) is generated from **hydroelectricity** in 20

In [5]:
# Engine
engine = sq.create_engine('mysql+mysqlconnector://root:Malharsql@127.0.0.1/data604', echo=False)

#### 1. Total Renewable Energy Production by Country
- This query calculates the total renewable energy production for each country in 2023 and 2022.
- The result is sorted in descending order by the total renewable energy production in 2023/2022.
- Highlights the countries leading in total renewable energy production.


In [6]:
query = '''
SELECT Country, 
       (Wind_2023 + Solar_2023 + Hydro_2023 + OtherRenewables_2023) AS Total_Renewable_2023,
        (Wind_2022 + Solar_2022 + Hydro_2022 + OtherRenewables_2022) AS Total_Renewable_2022
FROM renewable_energy
ORDER BY Total_Renewable_2023 DESC;
'''
# Ordering by 2023 totals in descending order to identify leaders


df_temp = pd.read_sql(query, engine)
print(df_temp.head())
del df_temp

  Country  Total_Renewable_2023  Total_Renewable_2022
0   China           2894.140732           2670.587265
1      US            973.667343            968.617340
2  Brazil            631.476933            593.904800
3  Canada            421.158884            452.275186
4   India            381.964584            378.705719


#### 2. Contribution of Renewable Energy Types in Canada (2023)
- Which renewable type (wind, solar, hydro, others) contributes to Canada’s total renewable energy in 2023.
- Highlights the Canada’s dominant renewable energy source in 2023.
- I will use this to compare the report which mentions that more than half of the electricity in Canada (60%) is generated from hydroelectricity in 2021.


In [7]:
# Query
query = '''
SELECT Country,
       Wind_2023/(Wind_2023 + Solar_2023 + Hydro_2023 + OtherRenewables_2023)*100 AS Wind_Contribution, 
       Solar_2023/(Wind_2023 + Solar_2023 + Hydro_2023 + OtherRenewables_2023)*100  AS Solar_Contribution,
       Hydro_2023/(Wind_2023 + Solar_2023 + Hydro_2023 + OtherRenewables_2023)*100  AS Hydro_Contribution, 
       OtherRenewables_2023/(Wind_2023 + Solar_2023 + Hydro_2023 + OtherRenewables_2023)*100  AS Other_Contribution
FROM renewable_energy
WHERE Country = 'Canada';
'''

df_temp = pd.read_sql(query, engine)
df_temp.head()

Country  Wind_Contribution  Solar_Contribution  Hydro_Contribution   
0  Canada           9.245037            1.814423           86.474848  \

   Other_Contribution  
0            2.465693

- From the above query, we can see that Canada's renewable energy mix in 2023 is 9.25% wind, 1.81% solar, 86.47% hydro, and 2.47% other renewables.
- We can also see that the contribution of hydro has increased from 2021 to 2023. This is because the hydro contribution in 2021 was 60% and in 2023 it is 86.47%.

#### 3. Percentage Change in Total Renewable Energy by Country
- Growth rate of total renewable energy production for each country from 2022 to 2023.
- We can identifies nations who are investing more in renewable energy production.

In [10]:
# Percentage Change in Total Renewable Energy by Country
query = '''
SELECT Country, 
       ((Wind_2023 + Solar_2023 + Hydro_2023 + OtherRenewables_2023) - 
        (Wind_2022 + Solar_2022 + Hydro_2022 + OtherRenewables_2022)) * 100.0 /
        (Wind_2022 + Solar_2022 + Hydro_2022 + OtherRenewables_2022) AS Percentage_Change
FROM renewable_energy
WHERE (Wind_2022 + Solar_2022 + Hydro_2022 + OtherRenewables_2022) > 0
ORDER BY Percentage_Change DESC;
'''

df_temp = pd.read_sql(query, engine)
df_temp.head(10)

Country  Percentage_Change
0                  Qatar         258.582080
1           Saudi Arabia         152.885988
2   United Arab Emirates          78.380510
3               Slovenia          56.495215
4                   Iran          55.127892
5        North Macedonia          36.193001
6                Croatia          33.611890
7             Uzbekistan          33.094284
8  Other Southern Africa          27.902717
9            Switzerland          26.960973

From the above query, we can see that the top three country with the highest percentage change in total renewable energy from 2022 to 2023 is:
- `Qatar`
- `Saudi Arabia`
- `United Arab Emirates`

What is interesting is that these countries are located in the Middle East which is rich in solar energy resources.
Let's see the contribution of renewable energy types in these countries in 2023.

### 4. Solar Energy Contribution in Qatar, UAE, and Saudi Arabia (2023)
- Contribution of renewable energy sources in countries like Qatar, UAE, and Saudi Arabia.
- Shows how these countries' renewable energy portfolios are shaped by their geographical location.

In [17]:
query = '''
SELECT Country,
       Wind_2023/(Wind_2023 + Solar_2023 + Hydro_2023 + OtherRenewables_2023)*100 AS Wind_Contribution, 
       Solar_2023/(Wind_2023 + Solar_2023 + Hydro_2023 + OtherRenewables_2023)*100  AS Solar_Contribution,
       Hydro_2023/(Wind_2023 + Solar_2023 + Hydro_2023 + OtherRenewables_2023)*100  AS Hydro_Contribution, 
       OtherRenewables_2023/(Wind_2023 + Solar_2023 + Hydro_2023 + OtherRenewables_2023)*100  AS Other_Contribution
FROM renewable_energy
WHERE Country IN ('Qatar', 'Saudi Arabia', 'United Arab Emirates')
ORDER BY Solar_Contribution DESC;
'''

df_temp = pd.read_sql(query, engine)
df_temp.head(10)

# As expected, these countries have a high contribution of solar energy to the total renewable energy in 2023.

Country  Wind_Contribution  Solar_Contribution   
0  United Arab Emirates           0.389944           99.369958  \
1                 Qatar           0.000000           90.762137   
2          Saudi Arabia          25.133279           74.866721   

   Hydro_Contribution  Other_Contribution  
0                 0.0            0.240097  
1                 0.0            9.237863  
2                 0.0            0.000000

As expected, these countries have a high contribution of solar energy to the total renewable energy in 2023.

#### 5. Ranking Countries by Renewable Energy Types

- Ranks countries based on each type of renewable energy in 2023.
- Highlights contries dominating in specific renewable energy categories.

In [14]:

query = '''
SELECT Country, 
       Wind_2023, 
       RANK() OVER (ORDER BY Wind_2023 DESC) AS Wind_Rank,
       Solar_2023, 
       RANK() OVER (ORDER BY Solar_2023 DESC) AS Solar_Rank,
       Hydro_2023, 
       RANK() OVER (ORDER BY Hydro_2023 DESC) AS Hydro_Rank,
       OtherRenewables_2023, 
       RANK() OVER (ORDER BY OtherRenewables_2023 DESC) AS Other_Rank
FROM renewable_energy
ORDER BY Solar_2023 DESC;
'''

df_temp = pd.read_sql(query, engine)
df_temp.head(23)

Country   Wind_2023  Wind_Rank  Solar_2023  Solar_Rank   
0                  China  885.870000          1   584.15000           1  \
1                     US  429.531000          2   240.52500           2   
2                  India   82.106400          5   113.40900           3   
3                  Japan   10.009000         27    96.99360           4   
4                Germany  142.103000          3    61.21600           5   
5                 Brazil   95.508900          4    51.48210           6   
6                  Spain   64.227500          7    46.84730           7   
7              Australia   31.867900         12    44.99430           8   
8                  Italy   23.530400         15    31.22740           9   
9            South Korea    3.389800         40    29.36860          10   
10                Mexico   21.651000         16    27.14330          11   
11               Vietnam   11.367000         25    25.70200          12   
12                France   52.311300          8    22.24760          13   
13           Netherlands   28.885000         13    21.17300          14   
14                Turkey   34.069700         11    20.51890          15   
15                 Chile    9.750720         28    18.17770          16   
16        United Kingdom   81.989500          6    13.82650          17   
17  United Arab Emirates    0.053951         74    13.74840          18   
18                Taiwan    6.201350         32    12.90870          19   
19                Poland   24.130000         14    11.66880          20   
20                Greece   11.013700         26     8.49951          21   
21                Canada   38.936300          9     7.64160          22   
22                Israel    0.440931         62     7.51267          23   

     Hydro_2023  Hydro_Rank  OtherRenewables_2023  Other_Rank  
0   1226.000000           1            198.125000           1  
1    236.321000           4             67.290100           2  
2    149.167000           6             37.281600           7  
3     74.499300          11             42.045200           5  
4     19.639000          37             49.491000           4  
5    428.653000           2             55.832700           3  
6     25.475000          30              5.407510          27  
7     15.261200          39              3.125680          33  
8     38.923400          19             20.310300           9  
9      3.716420          57             19.246600          11  
10    20.399400          36              7.099760          23  
11    80.904000          10              0.853000          47  
12    55.522400          17              9.705620          17  
13     0.068000          77              8.036090          22  
14    63.854200          14             19.490100          10  
15    24.029700          34              6.027010          25  
16     5.194500          51             33.990000           8  
17     0.000000          83              0.033219          74  
18     3.963280          54              1.945630          37  
19     2.396320          60              8.246040          21  
20     3.947120          55              0.701957          50  
21   364.197000           3             10.384500          16  
22     0.023244          79              0.103054          71

In [28]:
query = '''
SELECT Country,
       Wind_2023/(Wind_2023 + Solar_2023 + Hydro_2023)*100 AS Wind_Contribution, 
       Solar_2023/(Wind_2023 + Solar_2023 + Hydro_2023 )*100  AS Solar_Contribution,
       Hydro_2023/(Wind_2023 + Solar_2023 + Hydro_2023 )*100  AS Hydro_Contribution
FROM renewable_energy
WHERE Country IN ('Canada', 'China', 'US', 'India', 'Germany', 'UK', 'France', 'Brazil', 'Japan', 'Australia', 'Russia', 'South Korea')
'''

df_temp = pd.read_sql(query, engine)
df_temp.head(20)

# As expected, these countries have a high contribution of solar energy to the total renewable energy in 2023.

Country  Wind_Contribution  Solar_Contribution  Hydro_Contribution
0       Canada           9.478754            1.860292           88.660954
1           US          47.389837           26.536992           26.073171
2       Brazil          16.591660            8.943389           74.464951
3       France          40.214304           17.102856           42.682840
4      Germany          63.735322           27.456292            8.808386
5    Australia          34.592627           48.841374           16.565999
6        China          32.858490           21.667161           45.474348
7        India          23.820842           32.902503           43.276654
8        Japan           5.514535           53.439445           41.046020
9  South Korea           9.293532           80.517451           10.189018

### REFERENCES:

The Sustainable Choice. (n.d.). Making sense of energy units, from https://thesustainablechoice.com/making-sense-of-energy-units/

Canada Energy Regulator. (n.d.). Provincial and territorial energy profiles – Canada, from https://www.cer-rec.gc.ca/en/data-analysis/energy-markets/provincial-territorial-energy-profiles/provincial-territorial-energy-profiles-canada.html

